overlord/snapstate/autorefresh_gating.go

```diff
@@ -631,3 +632,55 @@ var snapsToRefresh = func(gatingTask *state.Task) ([]*refreshCandidate, error) {
 
 	return candidates, nil
 }
+
+// AutoRefreshForGatingSnap triggers an auto-refresh change for all
+// snaps held by the given gating snap. This should only be called if the
+// gate-auto-refresh-hook feature is enabled.
+// TODO: this should be restricted as it doesn't take refresh timer/refresh hold
+// into account.
+func AutoRefreshForGatingSnap(st *state.State, gatingSnap string) error {
+	// ensure nothing is in flight already
+	if autoRefreshInFlight(st) {
+		return fmt.Errorf("there is an auto-refresh in progress")
+	}
+
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+
+	var hasHeld bool
+	for _, holdingSnaps := range gating {
+		if _, ok := holdingSnaps[gatingSnap]; ok {
+			hasHeld = true
+			break
+		}
+	}
+	if !hasHeld {
+		return fmt.Errorf("no snaps are held by snap %q", gatingSnap)
+	}
+
+	// NOTE: this will unlock and re-lock state for network ops
+	// XXX: should we refresh assertions (just call AutoRefresh()?)
```

Comment:
yes, we should do something about refreshing assertions, mainly for the usual reasons which is validations. We might want to do that separately though because I think the work on validation-sets will also need to tweak how we organize refreshing assertions